### PR TITLE
GDPR-110 Finding offenders with duplicate PNCs

### PR DESCRIPTION
**Pull request by brightonsbox**

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/nomis/datacompliance/repository/jpa/repository/DuplicateOffenderRepository.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+     * into account that either 2 or 4 digits may be used in
+     * the first part (year) of the number.
+     */
+    @Query(value =
+                    "SELECT O1.OFFENDER_ID_DISPLAY FROM OFFENDERS O1 " +
+                    "INNER JOIN OFFENDER_IDENTIFIERS OI1 " +
+                    "ON O1.OFFENDER_ID = OI1.OFFENDER_ID " +
+                    "AND OI1.IDENTIFIER_TYPE = 'PNC'" +
+                    "AND REGEXP_SUBSTR(UPPER(OI1.IDENTIFIER), '[0-9]{2,2}/[0-9]+[A-Z]+') IN (" +
+                    "    SELECT REGEXP_SUBSTR(UPPER(OI2.IDENTIFIER), '[0-9]{2,2}/[0-9]+[A-Z]+') FROM OFFENDERS O2" +
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/hmpps/nomis/datacompliance/repository/jpa/repository/DuplicateOffenderRepository.java:26]: Not sure what the regex does here, but I guess it copes with `2020/001234X` matching `2020/001234x` does it cope with 
`2019/001234X` matching `19/001234X` or 
`2019/001234X` matching `2019/1234X` or
`19/001234X` matching `2019/1234x` ?

Comment (andymarke) [reply src/main/java/uk/gov/justice/hmpps/nomis/datacompliance/repository/jpa/repository/DuplicateOffenderRepository.java:26]: Though not sure how NOMIS pnc numbers are constrained so not sure what variations are in NOMIS.

Comment (andymarke) [reply src/main/java/uk/gov/justice/hmpps/nomis/datacompliance/repository/jpa/repository/DuplicateOffenderRepository.java:26]: I just saw test, so it is the zero padding that might need a test assuming the regex works?

File: src/main/java/uk/gov/justice/hmpps/nomis/datacompliance/repository/jpa/model/DuplicateOffender.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.Table;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Entity
+@Table(name = "OFFENDERS")
+public class DuplicateOffender {
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/hmpps/nomis/datacompliance/repository/jpa/model/DuplicateOffender.java:17]: I suspect you can do the query below without defining this table entity (assuming there is a OFFENDER jpa model already)

File: src/main/java/uk/gov/justice/hmpps/nomis/datacompliance/repository/jpa/repository/DuplicateOffenderRepository.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.CrudRepository;
+import org.springframework.stereotype.Repository;
+import uk.gov.justice.hmpps.nomis.datacompliance.repository.jpa.model.DuplicateOffender;
+
+import java.util.List;
+
+@Repository
+public interface DuplicateOffenderRepository extends CrudRepository<DuplicateOffender, String> {
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/hmpps/nomis/datacompliance/repository/jpa/repository/DuplicateOffenderRepository.java:12]: Maybe this doesn't need extend CrudRepostory if this is all it does?

File: src/main/java/uk/gov/justice/hmpps/nomis/datacompliance/repository/jpa/repository/DuplicateOffenderRepository.java (hunk 1 above)
Comment (brightonsbox) [reply src/main/java/uk/gov/justice/hmpps/nomis/datacompliance/repository/jpa/repository/DuplicateOffenderRepository.java:26]: It wasn't coping with zero padding, I'll add that in, thanks for flagging.

**Pull request merged**